Feature/knn

Rusty89 opened the pull request:
Made KNN, EditedKNN, CondensedKNN, loss functions and mode and mean functions, deleted bunch of useless classes

Rusty89 committed:
Removed unecessary classes

added new Algorithms class, added some new math functions i need, wrote most of KNN

Rusty89 committed:
Mean squared error added

Rusty89 committed:
Added more functionality

Fixed an indicing error where i was including the classifcation itself

Rusty89 committed:
Added loss functions

added loss functions to by making a confusion matrix and RMSE and MAE

Rusty89 committed:
Comments added

Rusty89 committed:
Fixed a scaling error

Needed to normalize to the orignal set, not the test and training set to get proper error output

Rusty89 committed:
Added test loops

Added some test loops to play with, fixed a NaN issue in the confusion matrix

Rusty89 committed:
KNN edited added

Edited KNN iteratively removes each bad point from the set, not collectively, maybe we will change later if you guys want

Rusty89 committed:
Condensed KNN added

added condensed KNN

Rusty89 committed:
Driver changes

Rusty89 committed:
EditedKnn

Changed edited to finish quicker, has to improve by 5% in each metric

RialJohnson commented on Project 2/k-NNAlgorithms/src/Algorithms.java:
```diff
@@ -0,0 +1,171 @@
+    //base KNN function
+    public static ArrayList<String> KNN(ArrayList<ArrayList<String>>trainingData, ArrayList<ArrayList<String>>testingData, int k, boolean regression, boolean euclidean){
+        //variables to hold sizes of things
+        int lengthOfTrainingSet = trainingData.size();
```
Doesn't really matter in our case, but probably better practice to use a const/readonly instead of int.

RialJohnson commented on Project 2/k-NNAlgorithms/src/Algorithms.java:
```diff
@@ -0,0 +1,171 @@
+        ArrayList<ArrayList<String>> finalEditedTrainingData =new ArrayList<>();
+
+
+        while(currentAccuracy>prevAccuracy && currentPrecision>prevPrecision && currentRecall>prevRecall){//while edited set is still improving
```
It's probably best to set a variable name for a condition like this

RialJohnson commented on Project 2/k-NNAlgorithms/src/Algorithms.java:
```diff
@@ -0,0 +1,171 @@
+                samplePoint.add(editedTrainingData.get(i));//give it the value of editedTraining data at i
+                ArrayList<String> classification = KNN(editedTrainingData,samplePoint, k, regression,euclidean);
+
+                if(classification.get(0).equals(samplePoint.get(0).get(indexOfClassification))){ //sample point classified correctly
```
0 can be a const called classifyingColumn or index or something.

RialJohnson commented on Project 2/k-NNAlgorithms/src/Algorithms.java:
```diff
@@ -0,0 +1,171 @@
+
+            //set prevs and currs to determine if a decrease in accuracy,precision or recall has
+            //occurred, exit while loop if so.
+            prevAccuracy= currentAccuracy+.05;
```
Just a note that we should tune this and comment why we're adding to this additional factor.

RialJohnson commented on Project 2/k-NNAlgorithms/src/Algorithms.java:
```diff
@@ -0,0 +1,171 @@
+
+                    //add the point to condensed training data if it's classification doesn't equal the classification of the point that
+                    //is min distance from it, and it doesn't already exist in condensed
+                    if(condensedTrainingData.get(indexOfMinValue).get(indexOfClassification).equals(trainingData.get(i).get(indexOfClassification))){
```
This logic block can probably be rewritten such that we don't need paths that don't execute any code (negate the condition or something). Also another potential candidate for giving a name to the condition since it's so long.

Rusty89 replied:
I agree, I was being somewhat sloppy this morning

RialJohnson commented on Project 2/k-NNAlgorithms/src/Driver.java:
```diff
@@ -1,16 +1,56 @@
 
+        //test loop for regression data
+
+        for (int k = 1; k <6 ; k+=2) {
```
why is k < 6 and incrementing by 2?

Rusty89 replied:
1,3,5 for K values

Rusty89 replied:
K values have to be odd

RialJohnson commented on Project 2/k-NNAlgorithms/src/Driver.java:
```diff
@@ -1,16 +1,56 @@
+        for (int k = 1; k <6 ; k+=2) {
+            double RMSE=0;
+            double absError=0;
+            for (int i = 0; i <10 ; i++) {
```
why is i < 10?

Rusty89 replied:
For the 10fold cross validation,

RialJohnson commented on Project 2/k-NNAlgorithms/src/MathFunction.java:
```diff
@@ -27,4 +27,176 @@ public static double hammingDistance(ArrayList<String> in1, ArrayList<String> in
+        ArrayList<ArrayList<String>> confusionMatrix= new ArrayList<>();
+        confusionMatrix.add(new ArrayList<>());
+        for (int i = 0; i <lengthOfData ; i++) {
+            if(!confusionMatrix.get(0).contains(testData.get(i).get(classIndex))){
```
Constant to clarify that 0 is used for classification column

Rusty89 replied:
It's a row, but yes that would be good

RialJohnson commented on Project 2/k-NNAlgorithms/src/MathFunction.java:
```diff
@@ -27,4 +27,176 @@ public static double hammingDistance(ArrayList<String> in1, ArrayList<String> in
+            falseNeg-=truePos;//get rid of true pos value in these results
+            falsePos-=truePos;
+
+            //calculates the precision and recall for each class
```
Can we use block comments for comments that are >2 lines?

Rusty89 replied:
Yes, I will do that from now on.

RialJohnson reviewed (changes requested):
I'll go ahead and make these changes, but you should read through them.

Rusty89 committed:
Logic Error

Forgot to update currMax in the mode function

Rusty89 committed:
Fixed major error in Edited

Edited had a major error where it would never actuall edit the final set, fixed it now

Rusty89 committed:
Made while loop condition a boolean

made the while loop condition cleaner

RialJohnson committed:
changed file structure in Driver

RialJohnson committed:
Merge branch 'feature/knn' of https://github.com/Rusty89/Machine-Learning-Projects into feature/knn

RialJohnson committed:
Cleaned up driver, push to fix merge conflicts

RialJohnson committed:
A lot of code cleanup

NickHager reviewed (approved):
+1
